File: Cloud Mail Samples/Cloud Mailer/cpp/cloudmailer.cpp
/*
 * Cloud Mail 2022 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of Cloud Mail in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/cloudmail
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

#include <iostream>
#include <string>
#include "../../include/cloudmail.h"

#define LINE_LEN 80
#define MESSAGE_LEN 1024


int main()
{
    char command[LINE_LEN];     // user's command

    CloudMailer cloudmailer1;

    // Welcome Message
    printf("Welcome to the Cloud Mailer demo.\n");
    printf("Follow the prompts to send an email using the component.\n");
    printf("------------------------------------------------------------\n");

    printf("Which mail service provider would you like to use to send your message? \n  [0] - Amazon SES\n  [1] - Gmail\n  [2] - Office365\n");
    fgets(command, LINE_LEN, stdin);
    if (std::stoi(command) == 0) {
        cloudmailer1.SetServiceProvider(0);
        printf("Enter your Amazon secret key: ");
        fgets(command, LINE_LEN, stdin);
        cloudmailer1.SetAccountAmazonAccessKey(command);
        printf("Enter your Amazon secret key: ");
        cloudmailer1.SetAccountAmazonSecretKey(command);
        fgets(command, LINE_LEN, stdin);
    }
    else if (std::stoi(command) == 1) {
        cloudmailer1.SetServiceProvider(1);
        cloudmailer1.SetOAuthServerAuthURL("https://accounts.google.com/o/oauth2/auth");
        cloudmailer1.SetOAuthServerTokenURL("https://accounts.google.com/o/oauth2/token");
        printf("Enter your OAuth Client ID: ");
        fgets(command, LINE_LEN, stdin);
        cloudmailer1.SetOAuthClientId(command);
        printf("Enter your OAuth client secret: ");
        fgets(command, LINE_LEN, stdin);
        cloudmailer1.SetOAuthClientSecret(command);
        printf("Enter desired authorization scopes (leave a space between each one if providing multiple): ");
        fgets(command, LINE_LEN, stdin);
        cloudmailer1.SetOAuthAuthorizationScope(command);
        printf("Authenticating...\n");
        cloudmailer1.Authorize();
        printf("Authentication successful.\n");
    }
    else if (std::stoi(command) == 2) {
        cloudmailer1.SetServiceProvider(2);
        cloudmailer1.SetOAuthServerAuthURL("https://login.microsoftonline.com/common/oauth2/v2.0/authorize");
        cloudmailer1.SetOAuthServerTokenURL("https://login.microsoftonline.com/common/oauth2/v2.0/token");
        printf("Enter your OAuth Client ID: ");
        fgets(command, LINE_LEN, stdin);
        cloudmailer1.SetOAuthClientId(command);
        printf("Enter your OAuth client secret: ");
        fgets(command, LINE_LEN, stdin);
        cloudmailer1.SetOAuthClientSecret(command);
        printf("Enter desired authorization scopes (leave a space between each one if providing multiple): ");
        fgets(command, LINE_LEN, stdin);
        cloudmailer1.SetOAuthAuthorizationScope(command);
        printf("Authenticating...\n");
        cloudmailer1.Authorize();
        printf("Authentication successful.\n");
    }
    else {
        throw std::invalid_argument("Invalid mail service provider.");
    }

    // Prompt for message information.
    printf("Provide email address(es) to receive message (leave a comma between each one if providing multiple): ");
    fgets(command, LINE_LEN, stdin);
    cloudmailer1.SetSendTo(command);
    printf("Provide email address that is sending message: ");
    fgets(command, LINE_LEN, stdin);
    cloudmailer1.SetFrom(command);
    printf("Provide message subject (optional): ");
    fgets(command, LINE_LEN, stdin);
    cloudmailer1.SetSubject(command);
    printf("Provide HTML message text: ");
    fgets(command, MESSAGE_LEN, stdin);
    cloudmailer1.SetMessageHTML(command);
    printf("Would you like to add an attachment? \n[0] - No \n[1] - Yes\n");
    fgets(command, LINE_LEN, stdin);
    if (std::stoi(command) == 1) {
        printf("Enter full path of attachment to add: ");
        fgets(command, LINE_LEN, stdin);
        std::string path = command;
        size_t end = path.find_last_not_of(" \n\r\t\f\v");    
        path = (end == std::string::npos) ? "" : path.substr(0, end + 1);
        cloudmailer1.AddAttachment(path.c_str());
    }

    // Send email
    printf("Sending message...\n");
    if (cloudmailer1.Send() != 0) {
        printf("Error %d: %s", cloudmailer1.GetLastErrorCode(), cloudmailer1.GetLastError());
    }
    else {
        printf("Message sent successfully");
    }
    exit(0);
}

File: Cloud Mail Samples/Cloud Mailer/python/cloudmailer.py
# 
# Cloud Mail 2024 Python Edition - Sample Project
# 
# This sample project demonstrates the usage of Cloud Mail in a 
# simple, straightforward way. It is not intended to be a complete 
# application. Error handling and other checks are simplified for clarity.
# 
# www.nsoftware.com/cloudmail
# 
# This code is subject to the terms and conditions specified in the 
# corresponding product license agreement which outlines the authorized 
# usage and restrictions.
# 

import sys
import string
from cloudmail import *

input = sys.hexversion<0x03000000 and raw_input or input

def ensureArg(args, prompt, index):
  if len(args) <= index:
    while len(args) <= index:
      args.append(None)
    args[index] = input(prompt)
  elif args[index] is None:
    args[index] = input(prompt)


cloudmailer1 = CloudMailer()

def fireSSLServerAuthentication(e):
  e.accept = True

try:
  cloudmailer1.on_ssl_server_authentication = fireSSLServerAuthentication

  print("Which mail service provider would you like to use to send your message?")
  print(" [0] - Amazon SES")
  print(" [1] - Gmail")
  print(" [2] - Office 365")

  cloudmailer1.set_service_provider(int(input()))

  # Prompt for authentication information.
  if (cloudmailer1.get_service_provider() == 0):
    cloudmailer1.set_account_amazon_access_key(input("Enter your Amazon access key:  "))
    cloudmailer1.set_account_amazon_secret_key(input("Enter your Amazon secret key:  "))
  elif (cloudmailer1.get_service_provider() == 1):
    cloudmailer1.set_o_auth_server_auth_url("https://accounts.google.com/o/oauth2/auth")
    cloudmailer1.set_o_auth_server_token_url("https://accounts.google.com/o/oauth2/token")
    cloudmailer1.set_o_auth_client_id(input("Enter your OAuth client ID:  "))
    cloudmailer1.set_o_auth_client_secret(input("Enter your OAuth client secret:  "))
    cloudmailer1.set_o_auth_authorization_scope(input("Enter desired authorization scopes (leave a space between each one if providing multiple):  "))
    print("Authenticating... ", end = "")
    cloudmailer1.authorize()
    print(" done.")
  elif (cloudmailer1.get_service_provider() == 2):
    cloudmailer1.set_o_auth_server_auth_url("https://login.microsoftonline.com/common/oauth2/v2.0/authorize")
    cloudmailer1.set_o_auth_server_token_url("https://login.microsoftonline.com/common/oauth2/v2.0/token")
    cloudmailer1.set_o_auth_client_id(input("Enter your OAuth client ID:  "))
    cloudmailer1.set_o_auth_client_secret(input("Enter your OAuth client secret:  "))
    cloudmailer1.set_o_auth_authorization_scope(input("Enter desired authorization scopes (leave a space between each one if providing multiple):  "))
    print("Authenticating... ", end = "")
    cloudmailer1.authorize()
    print(" done.")

  # Prompt for message information.
  cloudmailer1.set_send_to(input("Provide email address(es) to receive message (leave a comma between each one if providing multiple):  "))
  cloudmailer1.set_from(input("Provide email address that is sending message:  "))
  cloudmailer1.set_subject(input("Provide message subject (optional):  "))
  cloudmailer1.set_message_html(input("Provide HTML message text:  "))
  print("Would you like to add an attachment?")
  print(" [0] - No")
  print(" [1] - Yes")
  attachmentPrompt = (int)(input())
  if (attachmentPrompt == 1):
    attachmentPath = input("Provide full path of attachment to add:  ")
    if attachmentPath != "":
      cloudmailer1.add_attachment(attachmentPath)

  # Send email.
  print("Sending message... ", end = "")
  cloudmailer1.send()
  print(" sent.")
except ValueError as e:
  print("Must supply a valid number.")
except Exception as e:
  print(e)

